File: advent-of-code/advent_of_code_2016/day_10/solution.rb
# Refactored to get rid of these class variables and improve overall design
# Each bot is initialized with id and factory object, so that he has access the other bots/outputs

class Factory
  VALUES = [17, 61]
  attr_reader :input, :bots, :outputs
  attr_accessor :looked_for_bot

  def initialize(input)
    @input = input
    @bots = {}
    @outputs = {}
    @looked_for_bot = nil
    create_bots
  end

  def output_solutions
    shuffle_chips
    puts looked_for_bot
    puts outputs[0] * outputs[1] * outputs[2]
  end

  private

  def create_bots
    input.scan(/bot \d+/).uniq.map do |bot_id|
      id = bot_id.scan(/\d+/)[0].to_i
      bots[id] = Bot.new(id, self)
    end
  end

  def shuffle_chips
    parse_initial_instructions
    parse_shuffling_instructions
  end

  def parse_initial_instructions
    instructions = input.split("\n").select { |string| string.match?(/value/) }

    instructions.each do |string|
      value, id = string.scan(/\d+/).map(&:to_i)
      bots[id].receive_chip(value)
    end
  end

  def parse_shuffling_instructions
    instructions = input.split("\n").select { |string| string.match?(/^bot/) }

    instructions.each do |string|
      id = string.scan(/\d+/).map(&:to_i)[0]
      bots[id].save_instruction(string)
    end
  end
end

class Bot
  attr_reader :id, :factory
  attr_accessor :chips, :give_low_to, :give_high_to

  def initialize(id, factory)
    @id = id
    @factory = factory
    @chips = []
    @give_low_to = nil
    @give_high_to = nil
  end

  def receive_chip(value)
    chips << value
    factory.looked_for_bot = id if chips.sort == Factory::VALUES
    pass_on_chips if ready_to_handle
  end

  def save_instruction(string)
    tokens = string.scan(/bot \d+|output \d+/)
    self.give_low_to = parse_instruction(tokens[1])
    self.give_high_to = parse_instruction(tokens[2])
    pass_on_chips if ready_to_handle
  end

  private

  def parse_instruction(string)
    receiver = string.scan(/bot|output/)[0]
    id = string.scan(/\d+/)[0].to_i
    { receiver: receiver, id: id }
  end

  def ready_to_handle
    chips.length == 2 && give_low_to && give_high_to
  end

  def pass_on_chips
    self.chips = chips.sort
    min, max = chips[0], chips[1]

    give_chip(min, give_low_to)
    give_chip(max, give_high_to)
  end

  def give_chip(value, instruction)
    id = instruction[:id]

    if instruction[:receiver] == 'bot'
      factory.bots[id].receive_chip(value)
    else
      factory.outputs[id] = value
    end
  end
end

input = File.read('input.txt').chomp

factory = Factory.new(input)
factory.output_solutions

File: advent-of-code/advent_of_code_2016/day_10/solution.js
/*
input:
  - list of instructions, separated by a new line
  - two formats:
    - 'value x goes to bot y': describe the initial state
    - 'bot x gives low to bot y and high to bot z': moving around of data

output:
  - an integer
  - the number of the bot responsible for comparing values 61 and 17

notes:
  - the bot can also place the 'value' in an output bin
  - the sequence if the 'shuffling' matters

ideas:
  - initialize all needed variables
  - first process all the initial instructions ('value x to bot y')

questions:
  - is there some kind of ordering of the input instructions?
    - yes, a bot 'gives' his chips only when he has two of them
*/

const lookedForValues = [61, 17]
var bots = {}
var outputs = {}
var lookedForBot;

function findBot(input) {
  let instructions = chomp(input).split('\n');
  let initialInstructions = instructions.filter((instruction) => /^value/.test(instruction));
  let shufflingInstructions = instructions.filter((instruction) => /^bot/.test(instruction));

  setStartingState(initialInstructions)
  shuffleChips(shufflingInstructions)

  console.log(lookedForBot); // part 1
  console.log(outputs[0] * outputs[1] * outputs[2]); // part 2
}

function setStartingState(initialInstructions) {
  initialInstructions.forEach((string) => {
    let parsedNumbers = string.match(/\d+/g).map(Number);
    let value = parsedNumbers[0];
    let bot = parsedNumbers[1];

    giveChip(bot, value);
  });
}

function shuffleChips(shufflingInstructions) {
  while (shufflingInstructions.length > 0) {
    let remainingInstructions = [];

    for (let i = 0; i < shufflingInstructions.length; i += 1) {
      let currentInstruction = shufflingInstructions[i];
      let numberTokens = currentInstruction.match(/\d+/g);
      let wordTokens = currentInstruction.match(/bot|output/g);
      let givingBot = numberTokens[0];

      if (!bots[givingBot] || bots[givingBot].length < 2) {
        remainingInstructions.push(currentInstruction);
        continue;
      }

      if (hasValues(givingBot)) lookedForBot = givingBot;

      let low = Math.min(...bots[givingBot]);
      let high = Math.max(...bots[givingBot]);

      processInstruction(low, wordTokens[1], numberTokens[1]);
      processInstruction(high, wordTokens[2], numberTokens[2]);
    }

    shufflingInstructions = remainingInstructions;
  }
}

function processInstruction(chip, receiver, id) {
  if (receiver === 'bot') {
    giveChip(id, chip);
  } else {
    outputs[id] = chip;
  }
}

function giveChip(bot, value) {
  bots[bot] ? bots[bot].push(value) : bots[bot] = [value];
}

function hasValues(bot) {
  let chips = bots[bot];
  return chips.includes(lookedForValues[0]) && chips.includes(lookedForValues[1]);
}

function chomp(text) {
  return text.replace(/(\n)+$/, '');
}

function fileRead(filename) {
  let contents = fs.readFileSync(filename);
  return contents.toString();
}

let fs = require('fs');
let input = fileRead('input.txt');
findBot(input);
// part 1 => 161
// part 2 => 133163
